lib/command.py

```diff
@@ -1,108 +1,113 @@
-            tell_player(other_player, f"{player.name} says: {message}")
+from game_data import rooms
+from lib.models.game_state import GameState
 
+from lib.models.player import Player
 
-@register_command
-def look(player: Player, params=None):
-    # store the player's current room
-    current_location = rooms[player.location]
-
-    # send the player back the description of their current room
-    tell_player(player, current_location['description'])
-
-    playershere = []
-    # go through every player in the game
-    for other_player in game.players.values():
-        # if they're in the same room as the player
-        if other_player.location == player.location:
-            # ... and they have a name to be shown
-            if other_player.name:
-                # add their name to the list
-                playershere.append(other_player.name)
-
-    # send player a message containing the list of players in the room
-    tell_player(player, "Players here: {}".format(", ".join(playershere)))
-
-    # send player a message containing the list of exits from this room
-    tell_player(player, "Exits are: {}".format(", ".join(current_location["exits"])))
-
-
-@register_command
-def go(player: Player, params):
-    # store the exit name   
-    ex = params.lower()
-
-    # store the player's current room
-    current_location = rooms[player.location]
-
-    # if the specified exit is found in the room's exits list
-    if ex in current_location["exits"]:
 
-        # go through all the players in the game
-        for other_player in game.players.values():
-            # if player is in the same room and isn't the player
-            # sending the command
-            if other_player.location == player.location and other_player.uuid != player.uuid:
-                # send them a message telling them that the player
-                # left the room
-                tell_player(other_player.uuid, f"{player.name} left via exit '{ex}'")
+class Commands(object):
+    def __init__(self, game_state: GameState):
+        self.game_state = game_state
+        # The register_command decorator was nicer, but it's harder to do inside a class
+        # we can look at abstracting it out again, but we need to decide whether we want
+        # state to be sent to every command if we do that, or manually register them so
+        # they can be class methods (or do some other type of metaprogramming to register
+        # them inside a class
+        self.commands = {
+            "quit": self.quit,
+            "say": self.say,
+            "help": self.help,
+            "look": self.look,
+            "go": self.go,
+        }
+
+    def execute_command(self, player, command, param):
+        print(self.commands)
+        if self.commands.get(command):
+            self.commands[command](player, param)
+        else:
+            self.game_state.tell_player(player, "Unknown command '{}'".format(command))
```

Comment:
Would much prefer this to be `tell_player()` without the need to go down through all these layers of objects. I don't know if it makes sense having the `tell_player()` method on a `GameState` object.

Reply:
The problem is, tell player needs to know about the same state somehow, so we can either have:

`GameState.tell_player(player, message)`

`tell_player(gamestate, player, message)`

`Player.tell(message)` <- with this one, we still needed to find the Player, so there's an implicit call to gamestate somewhere